statsmodels/regression/heckman.py

```diff
@@ -0,0 +1,663 @@
+        treated = np.asarray(~np.isnan(endog))
+
+        try:
+            endog_nomissing = endog.copy()
+            endog_nomissing[~treated] = -99999
+        except (TypeError, AttributeError):
+            endog_nomissing = [endog[i] if treated[i] else -99999 for i in range(len(treated))]
+
+        # create 1-D array that will be np.nan for every row of exog_select that has any missing
+        # values and a finite value otherwise for the call to super().__init__ so that it can
+        # strip out rows where exog_select has missing data if missing option is set
+
+        if np.asarray(exog_select).ndim==2:
+            exog_select_1dnan_placeholder = \
+                [np.nan if any(np.isnan(row)) else 1 for row in np.asarray(exog_select)]
+        else:  # assume ==1
+            exog_select_1dnan_placeholder = [np.nan if np.isnan(row) else 1 for row in np.asarray(exog_select)]
+
+        if pd.__name__ in type(endog).__module__:
+            exog_select_1dnan_placeholder = pd.Series(exog_select_1dnan_placeholder, index=endog.index)
+        else:
+            exog_select_1dnan_placeholder = np.array(exog_select_1dnan_placeholder)
+
+        # create array of sequential row positions so that rows of exog_select that have missing
+        # data can be identified after call to super().__init__
+        obsno = np.array(list(range(len(endog))))
+
+        # call super().__init__
+        super(Heckman, self).__init__(
+            endog_nomissing, exog=exog,
+            exog_select_1dnan_placeholder=exog_select_1dnan_placeholder, obsno=obsno,
+            treated=treated,
+            **kwargs)
+
+        # put np.nan back into endog for treated rows
+        self.endog = self.data.endog = \
+            np.asarray(
+                [self.endog[i] if self.treated[i] else np.nan for i in range(len(self.treated))]
+                )
+
+        # strip out rows stripped out by call to super().__init__ in Z variable
+        self.exog_select = np.asarray([np.asarray(exog_select)[obs] for obs in self.obsno])
+
+        # store variable names of exog_select
+        try:
+            self.exog_select_names = exog_select.columns.tolist()
+        except AttributeError:
+            self.exog_select_names = None
+
+        # delete attributes created by the call to super().__init__ that are no longer needed
+        del self.exog_select_1dnan_placeholder
+        del self.obsno
+
+
+        # store observation counts
+        self.nobs_total = len(endog)
+        self.nobs_uncensored = self.nobs = np.sum(self.treated)
+        self.nobs_censored = self.nobs_total - self.nobs_uncensored
+
+
+    def initialize(self):
+        self.wendog = self.endog
+        self.wexog = self.exog
+
+
+    def whiten(self, data):
+        """
+        Model whitener for Heckman correction model does nothing.
+        """
+        return data
+
+
+    def get_datamats(self):
```

Comment:
these can be added as attributes, no need for extra methods and repeated evaluation